URLError in test_iers_a_dl_mirror (4.0rc1 - Windows 10)

Could be some local problem though.

```
Full Python Version:
3.7.5 (default, Oct 31 2019, 15:18:51) [MSC v.1916 64 bit (AMD64)]

encodings: sys: utf-8, locale: cp1252, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions:
Numpy: 1.17.3
Scipy: 1.3.1
Matplotlib: 3.1.1
h5py: not available
Pandas: not available
Cython: 0.29.14
Scikit-image: not available
asdf: not available
astropy-helpers: 4.0.rc1

Using Astropy options: remote_data: any.

plugins: arraydiff-0.3, astropy-header-0.1.1, doctestplus-0.5.0, openfiles-0.4.0, remotedata-0.3.2

________________________________________________ test_iers_a_dl_mirror ________________________________________________

self = <urllib.request.HTTPSHandler object at 0x0000018BEE72D448>, http_class = <class 'http.client.HTTPSConnection'>
req = <urllib.request.Request object at 0x0000018BF2EAC448>, http_conn_args = {'check_hostname': None, 'context': None}
host = 'datacenter.iers.org', h = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.

        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')

        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)

        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})

        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?

        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}

        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)

        try:
            try:
                h.request(req.get_method(), req.selector, req.data, headers,
>                         encode_chunked=req.has_header('Transfer-encoding'))

...\conda\envs\atest\lib\urllib\request.py:1317:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>, method = 'GET', url = '/data/9/finals2000A.all'
body = None
headers = {'Accept': '*/*', 'Connection': 'close', 'Host': 'datacenter.iers.org', 'User-Agent': 'astropy/iers'}

    def request(self, method, url, body=None, headers={}, *,
                encode_chunked=False):
        """Send a complete request to the server."""
>       self._send_request(method, url, body, headers, encode_chunked)

...\conda\envs\atest\lib\http\client.py:1252:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>, method = 'GET', url = '/data/9/finals2000A.all'
body = None
headers = {'Accept': '*/*', 'Connection': 'close', 'Host': 'datacenter.iers.org', 'User-Agent': 'astropy/iers'}
encode_chunked = False

    def _send_request(self, method, url, body, headers, encode_chunked):
        # Honor explicitly requested Host: and Accept-Encoding: headers.
        header_names = frozenset(k.lower() for k in headers)
        skips = {}
        if 'host' in header_names:
            skips['skip_host'] = 1
        if 'accept-encoding' in header_names:
            skips['skip_accept_encoding'] = 1

        self.putrequest(method, url, **skips)

        # chunked encoding will happen if HTTP/1.1 is used and either
        # the caller passes encode_chunked=True or the following
        # conditions hold:
        # 1. content-length has not been explicitly set
        # 2. the body is a file or iterable, but not a str or bytes-like
        # 3. Transfer-Encoding has NOT been explicitly set by the caller

        if 'content-length' not in header_names:
            # only chunk body if not explicitly set for backwards
            # compatibility, assuming the client code is already handling the
            # chunking
            if 'transfer-encoding' not in header_names:
                # if content-length cannot be automatically determined, fall
                # back to chunked encoding
                encode_chunked = False
                content_length = self._get_content_length(body, method)
                if content_length is None:
                    if body is not None:
                        if self.debuglevel > 0:
                            print('Unable to determine size of %r' % body)
                        encode_chunked = True
                        self.putheader('Transfer-Encoding', 'chunked')
                else:
                    self.putheader('Content-Length', str(content_length))
        else:
            encode_chunked = False

        for hdr, value in headers.items():
            self.putheader(hdr, value)
        if isinstance(body, str):
            # RFC 2616 Section 3.7.1 says that text default has a
            # default charset of iso-8859-1.
            body = _encode(body, 'body')
>       self.endheaders(body, encode_chunked=encode_chunked)

...\conda\envs\atest\lib\http\client.py:1298:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>, message_body = None

    def endheaders(self, message_body=None, *, encode_chunked=False):
        """Indicate that the last header line has been sent to the server.

        This method sends the request to the server.  The optional message_body
        argument can be used to pass a message body associated with the
        request.
        """
        if self.__state == _CS_REQ_STARTED:
            self.__state = _CS_REQ_SENT
        else:
            raise CannotSendHeader()
>       self._send_output(message_body, encode_chunked=encode_chunked)

...\conda\envs\atest\lib\http\client.py:1247:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>, message_body = None, encode_chunked = False

    def _send_output(self, message_body=None, encode_chunked=False):
        """Send the currently buffered request and clear the buffer.

        Appends an extra \\r\\n to the buffer.
        A message_body may be specified, to be appended to the request.
        """
        self._buffer.extend((b"", b""))
        msg = b"\r\n".join(self._buffer)
        del self._buffer[:]
>       self.send(msg)

...\conda\envs\atest\lib\http\client.py:1026:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>
data = b'GET /data/9/finals2000A.all HTTP/1.1\r\nAccept-Encoding: identity\r\nHost: datacenter.iers.org\r\nUser-Agent: astropy/iers\r\nAccept: */*\r\nConnection: close\r\n\r\n'

    def send(self, data):
        """Send `data' to the server.
        ``data`` can be a string object, a bytes object, an array object, a
        file-like object that supports a .read() method, or an iterable object.
        """

        if self.sock is None:
            if self.auto_open:
>               self.connect()

...\conda\envs\atest\lib\http\client.py:966:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>

    def connect(self):
        "Connect to a host on a given (SSL) port."

        super().connect()

        if self._tunnel_host:
            server_hostname = self._tunnel_host
        else:
            server_hostname = self.host

        self.sock = self._context.wrap_socket(self.sock,
>                                             server_hostname=server_hostname)

...\conda\envs\atest\lib\http\client.py:1422:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ssl.SSLContext object at 0x0000018BF31B1898>
sock = <socket.socket [closed] fd=-1, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
server_side = False, do_handshake_on_connect = True, suppress_ragged_eofs = True
server_hostname = 'datacenter.iers.org', session = None

    def wrap_socket(self, sock, server_side=False,
                    do_handshake_on_connect=True,
                    suppress_ragged_eofs=True,
                    server_hostname=None, session=None):
        # SSLSocket class handles server_hostname encoding before it calls
        # ctx._wrap_socket()
        return self.sslsocket_class._create(
            sock=sock,
            server_side=server_side,
            do_handshake_on_connect=do_handshake_on_connect,
            suppress_ragged_eofs=suppress_ragged_eofs,
            server_hostname=server_hostname,
            context=self,
>           session=session
        )

...\conda\envs\atest\lib\ssl.py:423:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'ssl.SSLSocket'>
sock = <socket.socket [closed] fd=-1, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
server_side = False, do_handshake_on_connect = True, suppress_ragged_eofs = True
server_hostname = 'datacenter.iers.org', context = <ssl.SSLContext object at 0x0000018BF31B1898>, session = None

    @classmethod
    def _create(cls, sock, server_side=False, do_handshake_on_connect=True,
                suppress_ragged_eofs=True, server_hostname=None,
                context=None, session=None):
        if sock.getsockopt(SOL_SOCKET, SO_TYPE) != SOCK_STREAM:
            raise NotImplementedError("only stream sockets are supported")
        if server_side:
            if server_hostname:
                raise ValueError("server_hostname can only be specified "
                                 "in client mode")
            if session is not None:
                raise ValueError("session can only be specified in "
                                 "client mode")
        if context.check_hostname and not server_hostname:
            raise ValueError("check_hostname requires server_hostname")

        kwargs = dict(
            family=sock.family, type=sock.type, proto=sock.proto,
            fileno=sock.fileno()
        )
        self = cls.__new__(cls, **kwargs)
        super(SSLSocket, self).__init__(**kwargs)
        self.settimeout(sock.gettimeout())
        sock.detach()

        self._context = context
        self._session = session
        self._closed = False
        self._sslobj = None
        self.server_side = server_side
        self.server_hostname = context._encode_hostname(server_hostname)
        self.do_handshake_on_connect = do_handshake_on_connect
        self.suppress_ragged_eofs = suppress_ragged_eofs

        # See if we are connected
        try:
            self.getpeername()
        except OSError as e:
            if e.errno != errno.ENOTCONN:
                raise
            connected = False
        else:
            connected = True

        self._connected = connected
        if connected:
            # create the SSL object
            try:
                self._sslobj = self._context._wrap_socket(
                    self, server_side, self.server_hostname,
                    owner=self, session=self._session,
                )
                if do_handshake_on_connect:
                    timeout = self.gettimeout()
                    if timeout == 0.0:
                        # non-blocking
                        raise ValueError("do_handshake_on_connect should not be specified for non-blocking sockets")
>                   self.do_handshake()

...\conda\envs\atest\lib\ssl.py:870:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ssl.SSLSocket [closed] fd=-1, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>, block = False

    @_sslcopydoc
    def do_handshake(self, block=False):
        self._check_connected()
        timeout = self.gettimeout()
        try:
            if timeout == 0.0 and block:
                self.settimeout(None)
>           self._sslobj.do_handshake()
E           ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)

...\conda\envs\atest\lib\ssl.py:1139: SSLCertVerificationError

During handling of the above exception, another exception occurred:

    @pytest.mark.remote_data
    def test_iers_a_dl_mirror():
>       iersa_tab = iers.IERS_A.open(iers.IERS_A_URL_MIRROR, cache=False)

...\conda\envs\atest\lib\site-packages\astropy\utils\iers\tests\test_iers.py:349:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
...\conda\envs\atest\lib\site-packages\astropy\utils\iers\iers.py:193: in open
    kwargs.update(file=download_file(file, cache=cache))
...\conda\envs\atest\lib\site-packages\astropy\utils\iers\iers.py:97: in download_file
    return utils.data.download_file(*args, **kwargs)
...\conda\envs\atest\lib\site-packages\astropy\utils\data.py:1175: in download_file
    raise errors[sources[0]]
...\conda\envs\atest\lib\site-packages\astropy\utils\data.py:1147: in download_file
    http_headers=http_headers)
...\conda\envs\atest\lib\site-packages\astropy\utils\data.py:979: in _download_file_from_source
    with urllib.request.urlopen(req, timeout=timeout) as remote:
...\conda\envs\atest\lib\urllib\request.py:222: in urlopen
    return opener.open(url, data, timeout)
...\conda\envs\atest\lib\urllib\request.py:525: in open
    response = self._open(req, data)
...\conda\envs\atest\lib\urllib\request.py:543: in _open
    '_open', req)
...\conda\envs\atest\lib\urllib\request.py:503: in _call_chain
    result = func(*args)
...\conda\envs\atest\lib\urllib\request.py:1360: in https_open
    context=self._context, check_hostname=self._check_hostname)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <urllib.request.HTTPSHandler object at 0x0000018BEE72D448>, http_class = <class 'http.client.HTTPSConnection'>
req = <urllib.request.Request object at 0x0000018BF2EAC448>, http_conn_args = {'check_hostname': None, 'context': None}
host = 'datacenter.iers.org', h = <http.client.HTTPSConnection object at 0x0000018BF5CCFF08>

    def do_open(self, http_class, req, **http_conn_args):
        """Return an HTTPResponse object for the request, using http_class.

        http_class must implement the HTTPConnection API from http.client.
        """
        host = req.host
        if not host:
            raise URLError('no host given')

        # will parse host:port
        h = http_class(host, timeout=req.timeout, **http_conn_args)
        h.set_debuglevel(self._debuglevel)

        headers = dict(req.unredirected_hdrs)
        headers.update({k: v for k, v in req.headers.items()
                        if k not in headers})

        # TODO(jhylton): Should this be redesigned to handle
        # persistent connections?

        # We want to make an HTTP/1.1 request, but the addinfourl
        # class isn't prepared to deal with a persistent connection.
        # It will try to read all remaining data from the socket,
        # which will block while the server waits for the next request.
        # So make sure the connection gets closed after the (only)
        # request.
        headers["Connection"] = "close"
        headers = {name.title(): val for name, val in headers.items()}

        if req._tunnel_host:
            tunnel_headers = {}
            proxy_auth_hdr = "Proxy-Authorization"
            if proxy_auth_hdr in headers:
                tunnel_headers[proxy_auth_hdr] = headers[proxy_auth_hdr]
                # Proxy-Authorization should not be sent to origin
                # server.
                del headers[proxy_auth_hdr]
            h.set_tunnel(req._tunnel_host, headers=tunnel_headers)

        try:
            try:
                h.request(req.get_method(), req.selector, req.data, headers,
                          encode_chunked=req.has_header('Transfer-encoding'))
            except OSError as err: # timeout error
>               raise URLError(err)
E               urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate in certificate chain (_ssl.c:1076)>

...\conda\envs\atest\lib\urllib\request.py:1319: URLError
```

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -1282,9 +1282,9 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
         downloading from HTTPS or TLS+FTP sources.  This can be used provide
         alternative paths to root CA certificates.  Additionally, if the key
         ``'certfile'`` and optionally ``'keyfile'`` and ``'password'`` are
-        included, they are passed to passed to a call to
-        `ssl.SSLContext.load_cert_chain`.  This can be used for performing
-        SSL/TLS client certificate authentication.
+        included, they are passed to `ssl.SSLContext.load_cert_chain`.  This
+        can be used for performing SSL/TLS client certificate authentication
+        for servers that require it.
 
     allow_insecure : bool, optional
         Allow downloading files over a TLS/SSL connection even when the server

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -78,10 +78,6 @@ class Conf(_config.ConfigNamespace):
         'Default User-Agent for HTTP request headers. This can be overwritten '
         'for a particular call via http_headers option, where available. '
         'This only provides the default value when not set by https_headers.')
-    default_allow_insecure = _config.ConfigItem(
-        False,
-        'Whether or not to allow insecure TLS/SSL connections to hosts with '
-        'unverifiable certificates.')
     remote_timeout = _config.ConfigItem(
         10.,
         'Time to wait for remote data queries (in seconds). Set this to zero '
@@ -1218,7 +1214,7 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
 
 def download_file(remote_url, cache=False, show_progress=True, timeout=None,
                   sources=None, pkgname='astropy', http_headers=None,
-                  ssl_context=None, allow_insecure=None):
+                  ssl_context=None, allow_insecure=False):
     """Downloads a URL and optionally caches the result.
 
     It returns the filename of a file containing the URL's contents.
@@ -1293,13 +1289,12 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
         Allow downloading files over a TLS/SSL connection even when the server
         certificate verification failed.  When set to `True` the potentially
         insecure download is allowed to proceed, but an
-        `~astropy.utils.exceptions.AstropyWarning` is issued.  The default
-        value for this can be set by
-        ``astropy.utils.data.conf.default_allow_insecure``.  If you are getting
-        frequently certificate verification warnings, consider installing or
-        upgrading `certifi`_ package, which provides frequently updated
-        certificates for common root CAs (i.e., a set similar to those used by
-        web browsers).  If installed, Astropy will use it automatically.
+        `~astropy.utils.exceptions.AstropyWarning` is issued.  If you are
+        frequently getting certificate verification warnings, consider
+        installing or upgrading `certifi`_ package, which provides frequently
+        updated certificates for common root CAs (i.e., a set similar to those
+        used by web browsers).  If installed, Astropy will use it
+        automatically.
 
         .. _certifi: https://pypi.org/project/certifi/
 
@@ -1329,8 +1324,6 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
     if http_headers is None:
         http_headers = {'User-Agent': conf.default_http_user_agent,
                         'Accept': '*/*'}
-    if allow_insecure is None:
-        allow_insecure = conf.default_allow_insecure
 
     missing_cache = ""
 

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -6,6 +6,7 @@ import atexit
 import contextlib
 import errno
 import fnmatch
+import functools
 import hashlib
 import os
 import io
@@ -23,7 +24,12 @@ import ftplib
 from tempfile import NamedTemporaryFile, gettempdir, TemporaryDirectory, mkdtemp
 from warnings import warn
 
-import certifi
+try:
+    import certifi
+except ImportError:
+    # certifi support is optional; when available it will be used for TLS/SSL
+    # downloads
+    certifi = None
 
 import astropy.config.paths
 from astropy import config as _config
@@ -72,6 +78,10 @@ class Conf(_config.ConfigNamespace):
         'Default User-Agent for HTTP request headers. This can be overwritten '
         'for a particular call via http_headers option, where available. '
         'This only provides the default value when not set by https_headers.')
+    default_allow_insecure = _config.ConfigItem(
+        False,
+        'Whether or not to allow insecure TLS/SSL connections to hosts with '
+        'unverifiable certificates.')
     remote_timeout = _config.ConfigItem(
         10.,
         'Time to wait for remote data queries (in seconds). Set this to zero '
@@ -1028,9 +1038,91 @@ class _FTPTLSHandler(urllib.request.FTPHandler):
                               persistent=False)
 
 
+@functools.lru_cache()
+def _build_urlopener(ftp_tls=False, ssl_context=None, allow_insecure=False):
+    """
+    Helper for building a `urllib.request.build_opener` which handles TLS/SSL.
+    """
+
+    ssl_context = dict(it for it in ssl_context) if ssl_context else {}
+    cert_chain = {}
+    if 'certfile' in ssl_context:
+        cert_chain.update({
+            'certfile': ssl_context.pop('certfile'),
+            'keyfile': ssl_context.pop('keyfile', None),
+            'password': ssl_context.pop('password', None)
+        })
+    elif 'password' in ssl_context or 'keyfile' in ssl_context:
+        raise ValueError(
+            "passing 'keyfile' or 'password' in the ssl_context argument "
+            "requires passing 'certfile' as well")
+
+    if 'cafile' not in ssl_context and certifi is not None:
+        ssl_context['cafile'] = certifi.where()
+
+    ssl_context = ssl.create_default_context(**ssl_context)
+
+    if allow_insecure:
+        ssl_context.check_hostname = False
+        ssl_context.verify_mode = ssl.CERT_NONE
+
+    if cert_chain:
+        ssl_context.load_cert_chain(**cert_chain)
+
+    https_handler = urllib.request.HTTPSHandler(context=ssl_context)
+
+    if ftp_tls:
+        urlopener = urllib.request.build_opener(_FTPTLSHandler(), https_handler)
+    else:
+        urlopener = urllib.request.build_opener(https_handler)
+
+    return urlopener
+
+
+def _try_url_open(source_url, timeout=None, http_headers=None, ftp_tls=False,
+                  ssl_context=None, allow_insecure=False):
+    """Helper for opening a URL while handling TLS/SSL verification issues."""
+
+    # Always try first with a secure connection
+    # _build_urlopener uses lru_cache, so the ssl_context argument must be
+    # converted to a hashshable type (a set of 2-tuples)
+    ssl_context = frozenset(ssl_context.items() if ssl_context else [])
+    urlopener = _build_urlopener(ftp_tls=ftp_tls, ssl_context=ssl_context,
+                                 allow_insecure=False)
+    req = urllib.request.Request(source_url, headers=http_headers)
+
+    try:
+        return urlopener.open(req, timeout=timeout)
+    except urllib.error.URLError as exc:
+        reason = exc.reason
+        if (isinstance(reason, ssl.SSLError)
+                and reason.reason == 'CERTIFICATE_VERIFY_FAILED'):
+            msg = (f'verification of TLS/SSL certificate at {source_url} '
+                   f'failed: this can mean either the server is '
+                   f'misconfigured or your local root CA certificates are '
+                   f'out-of-date; in the latter case this can usually be '
+                   f'addressed by installing the Python package "certifi" '
+                   f'(see the documentation for astropy.utils.data.download_url)')
+            if not allow_insecure:
+                msg += (f' or in both cases you can work around this by '
+                        f'passing allow_insecure=True, but only if you '
+                        f'understand the implications; the original error '
+                        f'was: {reason}')
+                raise urllib.error.URLError(msg)
+            else:
+                warn(msg, AstropyWarning)
+                # Try again with a new urlopener allowing insecure connections
+                urlopener = _build_urlopener(ftp_tls=ftp_tls,
+                        ssl_context=ssl_context, allow_insecure=True)
+                return urlopener.open(req, timeout=timeout)
+
+        raise
+
+
 def _download_file_from_source(source_url, show_progress=True, timeout=None,
                                remote_url=None, cache=False, pkgname='astropy',
-                               http_headers=None, ftp_tls=None):
+                               http_headers=None, ftp_tls=None,
+                               ssl_context=None, allow_insecure=False):
     from astropy.utils.console import ProgressBarOrSpinner
 
     if timeout == 0:
@@ -1069,16 +1161,9 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
             else:
                 raise
 
-    ssl_context = ssl.create_default_context(cafile=certifi.where())
-    https_handler = urllib.request.HTTPSHandler(context=ssl_context)
-
-    if ftp_tls:
-        urlopener = urllib.request.build_opener(_FTPTLSHandler(), https_handler)
-    else:
-        urlopener = urllib.request.build_opener(https_handler)
-
-    req = urllib.request.Request(source_url, headers=http_headers)
-    with urlopener.open(req, timeout=timeout) as remote:
+    with _try_url_open(source_url, timeout=timeout, http_headers=http_headers,
+            ftp_tls=ftp_tls, ssl_context=ssl_context,
+            allow_insecure=allow_insecure) as remote:
         info = remote.info()
         try:
             size = int(info['Content-Length'])
@@ -1132,7 +1217,8 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
 
 
 def download_file(remote_url, cache=False, show_progress=True, timeout=None,
-                  sources=None, pkgname='astropy', http_headers=None):
+                  sources=None, pkgname='astropy', http_headers=None,
+                  ssl_context=None, allow_insecure=None):
     """Downloads a URL and optionally caches the result.
 
     It returns the filename of a file containing the URL's contents.
@@ -1194,6 +1280,29 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
         ``User-Agent: some_value`` and ``Accept: */*``, where ``some_value``
         is set by ``astropy.utils.data.conf.default_http_user_agent``.
 
+    ssl_context : dict, optional
+        Keyword arguments to pass to `ssl.create_default_context` when
+        downloading from HTTPS or TLS+FTP sources.  This can be used provide
+        alternative paths to root CA certificates.  Additionally, if the key
+        ``'certfile'`` and optionally ``'keyfile'`` and ``'password'`` are
+        included, they are passed to passed to a call to
+        `ssl.SSLContext.load_cert_chain`.  This can be used for performing
+        SSL/TLS client certificate authentication.
+
+    allow_insecure : bool, optional
+        Allow downloading files over a TLS/SSL connection even when the server
+        certificate verification failed.  When set to `True` the potentially
+        insecure download is allowed to proceed, but an
+        `~astropy.utils.exceptions.AstropyWarning` is issued.  The default
+        value for this can be set by
+        ``astropy.utils.data.conf.default_allow_insecure``.  If you are getting
+        frequently certificate verification warnings, consider installing or
+        upgrading `certifi`_ package, which provides frequently updated
+        certificates for common root CAs (i.e., a set similar to those used by
+        web browsers).  If installed, Astropy will use it automatically.
+
+        .. _certifi: https://pypi.org/project/certifi/
+
     Returns
     -------
     local_path : str
@@ -1220,6 +1329,8 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
     if http_headers is None:
         http_headers = {'User-Agent': conf.default_http_user_agent,
                         'Accept': '*/*'}
+    if allow_insecure is None:
+        allow_insecure = conf.default_allow_insecure
 
     missing_cache = ""
 
@@ -1256,7 +1367,9 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None,
                     cache=cache,
                     remote_url=remote_url,
                     pkgname=pkgname,
-                    http_headers=http_headers)
+                    http_headers=http_headers,
+                    ssl_context=ssl_context,
+                    allow_insecure=allow_insecure)
             # Success!
             break
 

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -1071,12 +1071,16 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
             else:
                 raise
 
-    if ftp_tls:
-        urlopener = urllib.request.build_opener(_FTPTLSHandler())
+    if os.name == 'nt':
+        ssl_context = ssl.create_default_context(cafile=certifi.where())
     else:
-        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
-        ssl_context.load_verify_locations(certifi.where())
-        https_handler = urllib.request.HTTPSHandler(context=ssl_context)
+        ssl_context = ssl.create_default_context()
+
+    https_handler = urllib.request.HTTPSHandler(context=ssl_context)
+
+    if ftp_tls:
+        urlopener = urllib.request.build_opener(_FTPTLSHandler(), https_handler)
+    else:
         urlopener = urllib.request.build_opener(https_handler)
 
     req = urllib.request.Request(source_url, headers=http_headers)

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -23,16 +23,14 @@ import ftplib
 from tempfile import NamedTemporaryFile, gettempdir, TemporaryDirectory, mkdtemp
 from warnings import warn
 
+import certifi
+
 import astropy.config.paths
 from astropy import config as _config
 from astropy.utils.decorators import deprecated_renamed_argument
 from astropy.utils.exceptions import AstropyWarning
 from astropy.utils.introspection import find_current_module, resolve_name
 
-# Use certifi CA certificate store on Windows due to unreliable CA
-# certificate system store
-if os.name == 'nt':
-    import certifi
 
 # Order here determines order in the autosummary
 __all__ = [
@@ -1071,11 +1069,7 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
             else:
                 raise
 
-    if os.name == 'nt':
-        ssl_context = ssl.create_default_context(cafile=certifi.where())
-    else:
-        ssl_context = ssl.create_default_context()
-
+    ssl_context = ssl.create_default_context(cafile=certifi.where())
     https_handler = urllib.request.HTTPSHandler(context=ssl_context)
 
     if ftp_tls:

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -12,6 +12,7 @@ import io
 import re
 import shutil
 import socket
+import ssl
 import sys
 import urllib.request
 import urllib.error
@@ -27,6 +28,11 @@ from astropy import config as _config
 from astropy.utils.decorators import deprecated_renamed_argument
 from astropy.utils.exceptions import AstropyWarning
 from astropy.utils.introspection import find_current_module, resolve_name
+
+# Use certifi CA certificate store on Windows due to unreliable CA
+# certificate system store
+if os.name == 'nt':
+    import certifi
 
 # Order here determines order in the autosummary
 __all__ = [
@@ -1064,10 +1070,14 @@ def _download_file_from_source(source_url, show_progress=True, timeout=None,
                 ftp_tls = True
             else:
                 raise
+
     if ftp_tls:
         urlopener = urllib.request.build_opener(_FTPTLSHandler())
     else:
-        urlopener = urllib.request.build_opener()
+        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        ssl_context.load_verify_locations(certifi.where())
+        https_handler = urllib.request.HTTPSHandler(context=ssl_context)
+        urlopener = urllib.request.build_opener(https_handler)
 
     req = urllib.request.Request(source_url, headers=http_headers)
     with urlopener.open(req, timeout=timeout) as remote:

--- a/astropy/utils/data.py
+++ b/astropy/utils/data.py
@@ -1093,7 +1093,7 @@ def _try_url_open(source_url, timeout=None, http_headers=None, ftp_tls=False,
         reason = exc.reason
         if (isinstance(reason, ssl.SSLError)
                 and reason.reason == 'CERTIFICATE_VERIFY_FAILED'):
-            msg = (f'verification of TLS/SSL certificate at {source_url} '
+            msg = (f'Verification of TLS/SSL certificate at {source_url} '
                    f'failed: this can mean either the server is '
                    f'misconfigured or your local root CA certificates are '
                    f'out-of-date; in the latter case this can usually be '
@@ -1106,6 +1106,7 @@ def _try_url_open(source_url, timeout=None, http_headers=None, ftp_tls=False,
                         f'was: {reason}')
                 raise urllib.error.URLError(msg)
             else:
+                msg += '. Re-trying with allow_insecure=True.'
                 warn(msg, AstropyWarning)
                 # Try again with a new urlopener allowing insecure connections
                 urlopener = _build_urlopener(ftp_tls=ftp_tls,

--- a/astropy/utils/misc.py
+++ b/astropy/utils/misc.py
@@ -12,6 +12,7 @@ import inspect
 import json
 import os
 import signal
+import ssl
 import sys
 import traceback
 import unicodedata
@@ -24,6 +25,8 @@ from collections import defaultdict, OrderedDict
 
 from astropy.utils.decorators import deprecated
 
+if os.name == 'nt':
+    import certifi
 
 __all__ = ['isiterable', 'silence', 'format_exception', 'NumpyRNGContext',
            'find_api_page', 'is_path_hidden', 'walk_skip_hidden',
@@ -233,10 +236,15 @@ def find_api_page(obj, version=None, openinbrowser=True, timeout=None):
     req = urllib.request.Request(
         baseurl + 'objects.inv', headers={'User-Agent': f'Astropy/{version}'})
 
-    if timeout is None:
-        uf = urllib.request.urlopen(req)
+    if os.name == 'nt':
+        ssl_context = ssl.create_default_context(cafile=certifi.where())
     else:
-        uf = urllib.request.urlopen(req, timeout=timeout)
+        ssl_context = ssl.create_default_context()
+
+    if timeout is None:
+        uf = urllib.request.urlopen(req, context=ssl_context)
+    else:
+        uf = urllib.request.urlopen(req, timeout=timeout, context=ssl_context)
 
     try:
         oiread = uf.read()

--- a/astropy/utils/misc.py
+++ b/astropy/utils/misc.py
@@ -231,9 +231,9 @@ def find_api_page(obj, version=None, openinbrowser=True, timeout=None):
     # Custom request headers; see
     # https://github.com/astropy/astropy/issues/8990
     url = baseurl + 'objects.inv'
-    headers={'User-Agent': f'Astropy/{version}'}
+    headers = {'User-Agent': f'Astropy/{version}'}
     with get_readable_fileobj(url, encoding='binary', remote_timeout=timeout,
-            http_headers=headers) as uf:
+                              http_headers=headers) as uf:
         oiread = uf.read()
 
         # need to first read/remove the first four lines, which have info before

--- a/astropy/utils/misc.py
+++ b/astropy/utils/misc.py
@@ -12,7 +12,6 @@ import inspect
 import json
 import os
 import signal
-import ssl
 import sys
 import traceback
 import unicodedata
@@ -25,8 +24,6 @@ from collections import defaultdict, OrderedDict
 
 from astropy.utils.decorators import deprecated
 
-if os.name == 'nt':
-    import certifi
 
 __all__ = ['isiterable', 'silence', 'format_exception', 'NumpyRNGContext',
            'find_api_page', 'is_path_hidden', 'walk_skip_hidden',
@@ -201,8 +198,8 @@ def find_api_page(obj, version=None, openinbrowser=True, timeout=None):
 
     """
     import webbrowser
-    import urllib.request
     from zlib import decompress
+    from astropy.utils.data import get_readable_fileobj
 
     if (not isinstance(obj, str) and
             hasattr(obj, '__module__') and
@@ -233,20 +230,10 @@ def find_api_page(obj, version=None, openinbrowser=True, timeout=None):
 
     # Custom request headers; see
     # https://github.com/astropy/astropy/issues/8990
-    req = urllib.request.Request(
-        baseurl + 'objects.inv', headers={'User-Agent': f'Astropy/{version}'})
-
-    if os.name == 'nt':
-        ssl_context = ssl.create_default_context(cafile=certifi.where())
-    else:
-        ssl_context = ssl.create_default_context()
-
-    if timeout is None:
-        uf = urllib.request.urlopen(req, context=ssl_context)
-    else:
-        uf = urllib.request.urlopen(req, timeout=timeout, context=ssl_context)
-
-    try:
+    url = baseurl + 'objects.inv'
+    headers={'User-Agent': f'Astropy/{version}'}
+    with get_readable_fileobj(url, encoding='binary', remote_timeout=timeout,
+            http_headers=headers) as uf:
         oiread = uf.read()
 
         # need to first read/remove the first four lines, which have info before
@@ -266,8 +253,6 @@ def find_api_page(obj, version=None, openinbrowser=True, timeout=None):
                              'has changed?'.format(baseurl + 'objects.inv'))
 
         compressed = oiread[(idx+1):]
-    finally:
-        uf.close()
 
     decompressed = decompress(compressed).decode('utf-8')
 

--- a/astropy/utils/tests/test_data.py
+++ b/astropy/utils/tests/test_data.py
@@ -650,7 +650,7 @@ def test_download_certificate_verification_failed():
     # certificate verification error; we simulate this by passing a bogus
     # CA directory to the ssl_context argument
     ssl_context = {'cafile': None, 'capath': '/does/not/exist'}
-    msg = f'verification of TLS/SSL certificate at {TESTURL_SSL} failed'
+    msg = f'Verification of TLS/SSL certificate at {TESTURL_SSL} failed'
     with pytest.raises(urllib.error.URLError, match=msg):
         download_file(TESTURL_SSL, cache=False, ssl_context=ssl_context)
 

--- a/astropy/utils/tests/test_data.py
+++ b/astropy/utils/tests/test_data.py
@@ -29,6 +29,7 @@ import pytest
 from astropy import units as _u  # u is taken
 from astropy.config import paths
 import astropy.utils.data
+from astropy.utils.exceptions import AstropyWarning
 from astropy.utils.data import (
     CacheMissingWarning,
     CacheDamaged,
@@ -62,6 +63,7 @@ from astropy.utils.data import (
 CI = os.environ.get('CI', False) == "true"
 TESTURL = "http://www.astropy.org"
 TESTURL2 = "http://www.astropy.org/about.html"
+TESTURL_SSL = "https://www.astropy.org"
 TESTLOCAL = get_pkg_data_filename(os.path.join("data", "local.dat"))
 
 # NOTE: Python can be built without bz2 or lzma.
@@ -638,6 +640,28 @@ def test_download_cache():
     # Make sure lockdir was released
     lockdir = os.path.join(download_dir, "lock")
     assert not os.path.isdir(lockdir), "Cache dir lock was not released!"
+
+
+@pytest.mark.remote_data(source="astropy")
+def test_download_certificate_verification_failed():
+    """Tests for https://github.com/astropy/astropy/pull/10434"""
+
+    # First test the expected exception when download fails due to a
+    # certificate verification error; we simulate this by passing a bogus
+    # CA directory to the ssl_context argument
+    ssl_context = {'cafile': None, 'capath': '/does/not/exist'}
+    with pytest.raises(urllib.error.URLError) as exc:
+        download_file(TESTURL_SSL, cache=False, ssl_context=ssl_context)
+
+    msg = f'verification of TLS/SSL certificate at {TESTURL_SSL} failed'
+    assert msg in str(exc.value)
+
+    with pytest.warns(AstropyWarning) as warning_lines:
+        fnout = download_file(TESTURL_SSL, cache=False,
+                ssl_context=ssl_context, allow_insecure=True)
+
+    assert os.path.isfile(fnout)
+    assert len(warning_lines) == 1 and msg in str(warning_lines[0])
 
 
 def test_download_cache_after_clear(tmpdir, temp_cache, valid_urls):
@@ -1708,7 +1732,7 @@ def test_download_file_wrong_size(monkeypatch):
     def mockurl(remote_url, timeout=None):
         yield MockURL()
 
-    def mockurl_builder(tlscontext=None):
+    def mockurl_builder(*args, tlscontext=None, **kwargs):
         mock_opener = type('MockOpener', (object,), {})()
         mock_opener.open = mockurl
         return mock_opener
@@ -1723,7 +1747,7 @@ def test_download_file_wrong_size(monkeypatch):
         def read(self, length=None):
             return self.reader.read(length)
 
-    monkeypatch.setattr(urllib.request, "build_opener", mockurl_builder)
+    monkeypatch.setattr(astropy.utils.data, "_build_urlopener", mockurl_builder)
 
     with pytest.raises(urllib.error.ContentTooShortError):
         report_length = 1024

--- a/astropy/utils/tests/test_data.py
+++ b/astropy/utils/tests/test_data.py
@@ -650,18 +650,16 @@ def test_download_certificate_verification_failed():
     # certificate verification error; we simulate this by passing a bogus
     # CA directory to the ssl_context argument
     ssl_context = {'cafile': None, 'capath': '/does/not/exist'}
-    with pytest.raises(urllib.error.URLError) as exc:
+    msg = f'verification of TLS/SSL certificate at {TESTURL_SSL} failed'
+    with pytest.raises(urllib.error.URLError, match=msg):
         download_file(TESTURL_SSL, cache=False, ssl_context=ssl_context)
 
-    msg = f'verification of TLS/SSL certificate at {TESTURL_SSL} failed'
-    assert msg in str(exc.value)
-
-    with pytest.warns(AstropyWarning) as warning_lines:
+    with pytest.warns(AstropyWarning, match=msg) as warning_lines:
         fnout = download_file(TESTURL_SSL, cache=False,
                 ssl_context=ssl_context, allow_insecure=True)
 
+    assert len(warning_lines) == 1
     assert os.path.isfile(fnout)
-    assert len(warning_lines) == 1 and msg in str(warning_lines[0])
 
 
 def test_download_cache_after_clear(tmpdir, temp_cache, valid_urls):

--- a/astropy/utils/tests/test_misc.py
+++ b/astropy/utils/tests/test_misc.py
@@ -3,7 +3,7 @@
 import json
 import locale
 import os
-import socket
+import urllib.error
 from datetime import datetime
 
 import pytest
@@ -29,9 +29,10 @@ def test_signal_number_to_name_no_failure():
 @pytest.mark.remote_data
 def test_api_lookup():
     try:
-        strurl = misc.find_api_page('astropy.utils.misc', 'dev', False, timeout=3)
-        objurl = misc.find_api_page(misc, 'dev', False, timeout=3)
-    except socket.timeout:
+        strurl = misc.find_api_page('astropy.utils.misc', 'dev', False,
+                                    timeout=5)
+        objurl = misc.find_api_page(misc, 'dev', False, timeout=5)
+    except urllib.error.URLError:
         if os.environ.get('CI', False):
             pytest.xfail('Timed out in CI')
         else:

--- a/docs/changes/utils/10434.feature.rst
+++ b/docs/changes/utils/10434.feature.rst
@@ -1,0 +1,4 @@
+Added ``ssl_context`` and ``allow_insecure`` options to ``download_file``,
+as well as the ability to optionally use the ``certifi`` package to
+provide root CA certificates when downloading from sites secured with
+TLS/SSL.

--- a/docs/changes/utils/10434.feature.rst
+++ b/docs/changes/utils/10434.feature.rst
@@ -1,4 +1,3 @@
 Added ``ssl_context`` and ``allow_insecure`` options to ``download_file``,
-as well as the ability to optionally use the ``certifi`` package to
-provide root CA certificates when downloading from sites secured with
-TLS/SSL.
+as well as the ability to optionally use the ``certifi`` package to provide
+root CA certificates when downloading from sites secured with TLS/SSL.

--- a/docs/install.rst
+++ b/docs/install.rst
@@ -195,8 +195,9 @@ Requirements
 
 - `certifi <https://pypi.org/project/certifi/>`_: Useful when downloading
   files from HTTPS or FTP+TLS sites in case Python is not able to locate
-  up-to-date root CA certificates on your system; this package is already
-  included in many Python installations.
+  up-to-date root CA certificates on your system; this package is usually
+  already included in many Python installations (e.g. as a dependency of
+  the ``requests`` package).
 
 However, note that these packages require installation only if those particular
 features are needed. ``astropy`` will import even if these dependencies are not

--- a/docs/install.rst
+++ b/docs/install.rst
@@ -196,7 +196,7 @@ Requirements
 - `certifi <https://pypi.org/project/certifi/>`_: Useful when downloading
   files from HTTPS or FTP+TLS sites in case Python is not able to locate
   up-to-date root CA certificates on your system; this package is usually
-  already included in many Python installations (e.g. as a dependency of
+  already included in many Python installations (e.g., as a dependency of
   the ``requests`` package).
 
 However, note that these packages require installation only if those particular

--- a/docs/install.rst
+++ b/docs/install.rst
@@ -193,6 +193,11 @@ Requirements
   of sigma-clipping and other functionality that may require computing
   statistics on arrays with NaN values.
 
+- `certifi <https://pypi.org/project/certifi/>`_: Useful when downloading
+  files from HTTPS or FTP+TLS sites in case Python is not able to locate
+  up-to-date root CA certificates on your system; this package is already
+  included in many Python installations.
+
 However, note that these packages require installation only if those particular
 features are needed. ``astropy`` will import even if these dependencies are not
 installed.

--- a/docs/utils/data.rst
+++ b/docs/utils/data.rst
@@ -79,6 +79,23 @@ conveniently with the ``get_pkg_data_*`` functions::
    # these are all mappings from the name in sites.json (which is ASCII-only) to the "true" unicode names
    TUBITAK->TÜBİTAK
 
+
+.. note::
+
+    Sometimes when downloading files from internet resources secured with
+    TLS/SSL, you may get an exception regarding a certificate verification
+    error.  Typically this indicates that Python could not find an
+    up-to-date collection of `root certificates`_ on your system.  This is
+    especially common on Windows.  This problem can usually be resolved
+    by installing the `certifi`_ package, which Astropy will use if
+    available to verify remote connections.  In rare cases, certificate
+    verification may still fail if the remote server is misconfigured (e.g.,
+    with expired certificates).  In this case, you may pass the
+    ``allow_insecure=True`` argument to :func:`~astropy.utils.data.download_file`,
+    or add the setting ``astropy.utils.iers.conf.default_allow_insecure =
+    True`` in the :ref:`astropy config file <astropy_config>`.
+
+
 Usage From Outside Astropy
 ==========================
 
@@ -130,6 +147,8 @@ will be stored in the cache under the original URL requested::
    |========================================|  65M/ 65M (100.00%)        19s
 
 .. _Astropy data server: https://www.astropy.org/astropy-data/
+.. _root certificates: https://en.wikipedia.org/wiki/Root_certificate
+.. _certifi: https://pypi.org/project/certifi/
 
 Cache Management
 ================

--- a/docs/utils/data.rst
+++ b/docs/utils/data.rst
@@ -79,7 +79,6 @@ conveniently with the ``get_pkg_data_*`` functions::
    # these are all mappings from the name in sites.json (which is ASCII-only) to the "true" unicode names
    TUBITAK->TÜBİTAK
 
-
 .. note::
 
     Sometimes when downloading files from internet resources secured with
@@ -93,7 +92,8 @@ conveniently with the ``get_pkg_data_*`` functions::
     with expired certificates).  In this case, you may pass the
     ``allow_insecure=True`` argument to
     :func:`~astropy.utils.data.download_file` to allow the download with a
-    warning instead.
+    warning instead (not recommended unless you understand the `potential
+    risks <https://en.wikipedia.org/wiki/Man-in-the-middle_attack>`_).
 
 
 Usage From Outside Astropy

--- a/docs/utils/data.rst
+++ b/docs/utils/data.rst
@@ -91,9 +91,9 @@ conveniently with the ``get_pkg_data_*`` functions::
     available to verify remote connections.  In rare cases, certificate
     verification may still fail if the remote server is misconfigured (e.g.,
     with expired certificates).  In this case, you may pass the
-    ``allow_insecure=True`` argument to :func:`~astropy.utils.data.download_file`,
-    or add the setting ``astropy.utils.iers.conf.default_allow_insecure =
-    True`` in the :ref:`astropy config file <astropy_config>`.
+    ``allow_insecure=True`` argument to
+    :func:`~astropy.utils.data.download_file` to allow the download with a
+    warning instead.
 
 
 Usage From Outside Astropy

--- a/setup.cfg
+++ b/setup.cfg
@@ -64,6 +64,7 @@ test =
     packaging
 all =
     scipy>=1.1
+    certifi;platform_system=='Windows'
     dask[array]
     h5py
     beautifulsoup4

--- a/setup.cfg
+++ b/setup.cfg
@@ -64,7 +64,7 @@ test =
     packaging
 all =
     scipy>=1.1
-    certifi;platform_system=='Windows'
+    certifi
     dask[array]
     h5py
     beautifulsoup4

